core/src/Kokkos_Parallel_Reduce.hpp

```diff
@@ -91,8 +92,21 @@ struct Sum {
   KOKKOS_INLINE_FUNCTION
   void join(value_type& dest, const value_type& src) const { dest += src; }
 
+#if !KOKKOS_HALF_T_IS_FLOAT
   KOKKOS_INLINE_FUNCTION
-  void join(volatile value_type& dest, const volatile value_type& src) const {
+  void join(volatile Experimental::half_t& dest,
+            const volatile Experimental::half_t& src) const {
+    dest.atomic_add(src);
+  }
+#endif
```

Comment:
Oh my ...

Reply:
I am open to other options here. Perhaps a comment should be added here: if we try to invoke `dest += src` with the join operator below, the compiler throws a warning since `__half` does not have any volatile operators and this causes all the tests to fail since `-Werror` is used. Thoughts?

Reply:
This warning was related to having a return type of `half_t&` in the `+=` and `*=` volatile operators; that should be fixed in https://github.com/kokkos/kokkos/pull/4101/commits/67c343a8ee2e803697905456d2495fc01505cff0.